File: src/mind_os_builder/wiki/lint.py
from __future__ import annotations

import re
from dataclasses import asdict, dataclass, field
from pathlib import Path

import yaml


WIKILINK = re.compile(r"\[\[([^\]|#]+)(?:\|[^\]]+)?\]\]")
SYSTEM_FILES = {"index.md", "log.md", "lint-report.md"}
REQUIRED_FIELDS = {"domain", "sources", "created", "updated", "tags"}


@dataclass(frozen=True, slots=True)
class LintIssue:
    code: str
    path: str
    message: str
    level: str = "error"


@dataclass(slots=True)
class LintReport:
    issues: list[LintIssue] = field(default_factory=list)

    @property
    def error_count(self) -> int:
        return sum(issue.level == "error" for issue in self.issues)

    def to_dict(self) -> dict[str, object]:
        return {"error_count": self.error_count, "issues": [asdict(item) for item in self.issues]}


def _frontmatter(text: str) -> dict[str, object] | None:
    if not text.startswith("---\n"):
        return None
    marker = text.find("\n---\n", 4)
    if marker < 0:
        return None
    parsed = yaml.safe_load(text[4:marker])
    return parsed if isinstance(parsed, dict) else None


def lint_vault(root: Path) -> LintReport:
    report = LintReport()
    wiki = root / "wiki"
    if not wiki.is_dir():
        report.issues.append(LintIssue("wiki_missing", "wiki", "wiki directory is missing"))
        return report
    index_text = (wiki / "index.md").read_text(encoding="utf-8") if (wiki / "index.md").exists() else ""
    pages = [
        path
        for path in wiki.rglob("*.md")
        if path.name not in SYSTEM_FILES and "insights" not in path.relative_to(wiki).parts
    ]
    known = {path.stem for path in pages}
    inbound = {stem: 0 for stem in known}
    for page in pages:
        relative = page.relative_to(root).as_posix()
        text = page.read_text(encoding="utf-8")
        metadata = _frontmatter(text)
        if metadata is None:
            report.issues.append(LintIssue("frontmatter_missing", relative, "missing YAML frontmatter"))
        else:
            missing = sorted(REQUIRED_FIELDS - metadata.keys())
            if missing:
                report.issues.append(
                    LintIssue("frontmatter_incomplete", relative, f"missing fields: {', '.join(missing)}")
                )
        if len(text.splitlines()) > 500:
            report.issues.append(LintIssue("page_too_long", relative, "page exceeds 500 lines", "warning"))
        if f"[[{page.stem}]]" not in index_text:
            report.issues.append(LintIssue("index_missing", relative, "page is absent from wiki/index.md"))
        for target in WIKILINK.findall(text):
            stem = Path(target).stem
            if target.startswith(("raw/", "journals/")):
                continue
            if stem in known:
                inbound[stem] += 1
            else:
                report.issues.append(LintIssue("red_link", relative, f"missing target: {target}", "warning"))
    for stem, count in inbound.items():
        if count == 0 and f"[[{stem}]]" not in index_text:
            report.issues.append(LintIssue("orphan_page", stem, "page has no inbound links", "warning"))
    return report

File: src/wiki/lint.ts
import { lstat, readdir, readFile } from "node:fs/promises";
import { basename, dirname, join, relative as relativePath } from "node:path";
import { parse } from "yaml";
import { MindosError, resolveReadPath } from "../lib/paths.js";
import { blockedFromError, noopResult, previewResult, type CliResult } from "../lib/result.js";

const requiredFields = new Set(["domain", "sources", "created", "updated", "tags"]);
const systemFiles = new Set(["index.md", "log.md", "lint-report.md"]);
const wikilink = /\[\[([^\]|#]+)(?:\|[^\]]+)?\]\]/gu;

type Issue = { code: string; path: string; message: string; level: "error" | "warning" };

async function pagesAt(root: string, current = root): Promise<string[]> {
  const pages: string[] = [];
  for (const entry of await readdir(current, { withFileTypes: true })) {
    const path = join(current, entry.name);
    if (entry.isSymbolicLink()) {
      continue;
    }
    if (entry.isDirectory()) {
      if (entry.name !== "insights") {
        pages.push(...await pagesAt(root, path));
      }
    } else if (entry.isFile() && entry.name.endsWith(".md") && !systemFiles.has(entry.name)) {
      pages.push(path);
    }
  }
  return pages.sort((left, right) => left.localeCompare(right));
}

function frontmatter(content: string): Record<string, unknown> | undefined {
  if (!content.startsWith("---\n")) {
    return undefined;
  }
  const marker = content.indexOf("\n---\n", 4);
  if (marker < 0) {
    return undefined;
  }
  try {
    const parsed: unknown = parse(content.slice(4, marker));
    return typeof parsed === "object" && parsed !== null ? parsed as Record<string, unknown> : undefined;
  } catch {
    return undefined;
  }
}

export async function lintWiki(root: string): Promise<CliResult> {
  try {
    const wiki = await resolveReadPath(root, "wiki");
    if (!(await lstat(wiki)).isDirectory()) {
      throw new MindosError("mindos.filesystem.invalid_root", "vault is missing Wiki directory");
    }
    const indexPath = join(wiki, "index.md");
    const index = await readFile(indexPath, "utf8").catch(() => "");
    const pages = await pagesAt(wiki);
    const known = new Set(pages.map((path) => basename(path, ".md")));
    const inbound = new Map([...known].map((name) => [name, 0]));
    const issues: Issue[] = [];
    for (const page of pages) {
      const relative = relativePath(dirname(wiki), page).replaceAll("\\", "/");
      const content = await readFile(page, "utf8");
      const metadata = frontmatter(content);
      if (metadata === undefined) {
        issues.push({ code: "frontmatter_missing", path: relative, message: "missing YAML frontmatter", level: "error" });
      } else {
        const missing = [...requiredFields].filter((field) => !Object.hasOwn(metadata, field));
        if (missing.length > 0) {
          issues.push({ code: "frontmatter_incomplete", path: relative, message: `missing fields: ${missing.join(", ")}`, level: "error" });
        }
      }
      if (content.split("\n").length > 500) {
        issues.push({ code: "page_too_long", path: relative, message: "page exceeds 500 lines", level: "warning" });
      }
      const stem = basename(page, ".md");
      if (!index.includes(`[[${stem}]]`)) {
        issues.push({ code: "index_missing", path: relative, message: "page is absent from wiki/index.md", level: "error" });
      }
      for (const match of content.matchAll(wikilink)) {
        const target = match[1]?.trim() ?? "";
        if (target.startsWith("raw/") || target.startsWith("journals/")) {
          continue;
        }
        const targetStem = basename(target);
        if (known.has(targetStem)) {
          inbound.set(targetStem, (inbound.get(targetStem) ?? 0) + 1);
        } else {
          issues.push({ code: "red_link", path: relative, message: `missing target: ${target}`, level: "warning" });
        }
      }
    }
    for (const [stem, count] of inbound) {
      if (count === 0 && !index.includes(`[[${stem}]]`)) {
        issues.push({ code: "orphan_page", path: stem, message: "page has no inbound links", level: "warning" });
      }
    }
    const data = { error_count: issues.filter((issue) => issue.level === "error").length, issues, protected_paths: ["wiki/insights", "raw/logseq-import"] };
    return issues.length === 0 ? noopResult(data) : previewResult(data);
  } catch (error: unknown) {
    return blockedFromError(error, {});
  }
}
